common/src/main/java/org/apache/falcon/util/ResourceManager.java

```diff
@@ -0,0 +1,490 @@
+                public int compare(QueueElement o1, QueueElement o2) {
+                    String consumer1 = String.format("%s:%s", o1.status, o1.child);
+                    String consumer2 = String.format("%s:%s", o2.status, o2.child);
+                    return consumer1.compareTo(consumer2);
+                }
+            });
+        }
+
+        public void add(String child, String childPath, String zkNodeData, int version) {
+            String consumerId = getConsumerIdFromZnode(zkNodeData);
+            STATUS consumerStatus = getConsumerStatusFromZnode(zkNodeData);
+            q.add(new QueueElement(consumerId, child, childPath, consumerStatus, version));
+        }
+
+        public int findPosition(String zkNodePath) {
+            int pos = 0;
+            Iterator<ResourceQueue.QueueElement> itr = q.iterator();
+            while (itr.hasNext()) {
+                if (itr.next().childPath.equals(zkNodePath)) {
+                    break;
+                }
+                pos++;
+            }
+            return pos;
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        String zkHost = "c6402.ambari.apache.org";
+        String resourceName = "mysql-ds";
+        int resourceLimit = 3;
+        int numConsumers = 8;
+
+        setup(zkHost, resourceName, resourceLimit);
+        run(zkHost, resourceName, numConsumers);
+    }
+
+    public static void  setup(String zkHost, String resourceName, int resourceLimit) throws Exception {
+        ResourceManager rm = new ResourceManager(zkHost, resourceName, resourceLimit);
+        LOG.info("Creating a new Resource Manager with a resource Limit : " + resourceLimit);
+        rm.initialize();
+        LOG.info("Initialized Resource Manager");
+    }
+
+    public static void run(final String zkHosts, final String resourceName, int numConsumers) throws Exception {
+
+        ResourceManager rm = ResourceManager.get(zkHosts, resourceName);
+        LOG.info("Retrieved ResourceManager RM with max connection = {}", rm.resourceLimit);
+
+        final CountDownLatch lock = new CountDownLatch(numConsumers);
+        final Random random = new Random();
+        Thread[] threads = new Thread[numConsumers];
+        for (int i = 0; i < numConsumers; i++) {
+            threads[i] = new Thread("CONSUMER-" + i) {
+                @Override
+                public void run() {
+                    int delay = (5 + random.nextInt(16)) + 1000;
+                    try {
+                        ResourceManager rm = ResourceManager.get(zkHosts, resourceName);
+                        String znode = rm.getResource(this.getName());
+                        LOG.info("{} acquired lease znode {} work for {} seconds", getName(), znode, delay);
+                        Thread.currentThread().sleep(delay);
+                        znode = rm.releaseResource(getName());
+                        LOG.info("{} released znode {}", getName(), znode);
```

Comment:
Can you please help me understand the purpose of this block of code we get the resource ,wait  for some time and release the resource.